src/mumble/AudioOutput.cpp

```diff
@@ -168,35 +189,67 @@ void AudioOutput::addFrameToBuffer(ClientUser *sender, const Mumble::Protocol::A
 
 		qrwlOutputs.lockForWrite();
 
-		aop = new AudioOutputSpeech(sender, iMixerFreq, audioData.usedCodec, iBufferSize);
-		qmOutputs.replace(sender, aop);
+		speech = new AudioOutputSpeech(sender, iMixerFreq, audioData.usedCodec, iBufferSize);
+		qmOutputs.replace(sender, speech);
 	}
 
-	aop->addFrameToBuffer(audioData);
+	speech->addFrameToBuffer(audioData);
 
 	qrwlOutputs.unlock();
 }
 
-void AudioOutput::removeBuffer(const ClientUser *user) {
-	removeBuffer(qmOutputs.value(user));
+void AudioOutput::handleInvalidatedBuffer(AudioOutputBuffer *buffer) {
+	QWriteLocker locker(&qrwlOutputs);
+	for (auto iter = qmOutputs.begin(); iter != qmOutputs.end(); ++iter) {
+		if (iter.value() == buffer) {
+			qmOutputs.erase(iter);
+			delete buffer;
+			break;
+		}
+	}
 }
 
-void AudioOutput::removeBuffer(AudioOutputUser *aop) {
-	QWriteLocker locker(&qrwlOutputs);
-	QMultiHash< const ClientUser *, AudioOutputUser * >::iterator i;
-	for (i = qmOutputs.begin(); i != qmOutputs.end(); ++i) {
-		if (i.value() == aop) {
-			qmOutputs.erase(i);
-			delete aop;
+void AudioOutput::handlePositionedBuffer(AudioOutputBuffer *buffer, float x, float y, float z) {
+	QReadLocker locker(&qrwlOutputs);
```

Comment:
```suggestion
	QWriteLocker locker(&qrwlOutputs);
```
You are modifying contents inside `qmOutputs` - therefore you are not only reading in this function